### DISCUSSION TITLE
Populate: true does not initialize nested collections

### DISCUSSION THREAD

**Discussion by xndyz** [mikro-orm/mikro-orm#2869] [Q&A]
**Describe the bug**
`populate: true` option does not load nested properties. You must specify them individually in order to return the data.

**To Reproduce**

With a many to many entity relationship like so:

```
@Entity()
@ObjectType()
export class Member {
  @Field(() => [Chat], { nullable: true })
  @ManyToMany(() => Chat, 'members', { nullable: true, owner: true })
  chats = new Collection<Chat>(this);
}

@Entity()
@ObjectType()
export class Chat {
  @Field(() => [Member])
  @ManyToMany(() => Member, 'chats', { mappedBy: 'chats' })
  members = new Collection<Member>(this);
}
```

Querying the fields like so:

```
await this.em.findOneOrFail(
      Member,
      {id},
      {
        populate: true,
      },
    );
```
Returns:
```
Chat {
  id: '3a7e103c-fe89-4d72-a7e0-267f4031f2a8',
  status: 'open',
  members: Collection<Member> { initialized: false, dirty: false },
  messages: Collection<Message> { initialized: false, dirty: false }
```
But specifying the nested populate fields works:
```
await this.em.findOneOrFail(
      Member,
      {id},
      {
        populate: ['chat.members'],
      },
    );
```

**Expected behavior**
`populate: true` should return all nested collections.

**Comment by B4nan** [comment]
What version are you on?

 I could understand this behaviour in v4, but v5 should follow recursive relations too.

**Reply by xndyz** [reply]
I'm using the latest versions, which is why I suspect this is a bug:

```
"@mikro-orm/core": "^5.0.5",
    "@mikro-orm/migrations": "^5.0.5",
    "@mikro-orm/nestjs": "^5.0.0",
    "@mikro-orm/postgresql": "^5.0.5",
    "@mikro-orm/seeder": "^5.0.5",
```

**Reply by B4nan** [reply]
What queries does it fire? How does the ORM config look like?

**Reply by xndyz** [reply]
Config:

```
import { MikroOrmModuleSyncOptions } from '@mikro-orm/nestjs';

const {
  DATABASE_HOST,
  DATABASE_NAME,
  DATABASE_PORT,
  DATABASE_USER,
  DATABASE_PASS,
} = process.env;

const config: MikroOrmModuleSyncOptions = {
  type: 'postgresql',
  entities: ['./dist/src/**/*.entity.js'],
  entitiesTs: ['./src/**/*.entity.ts'],

  // DB CONNECTION
  dbName: DATABASE_NAME,
  host: DATABASE_HOST,
  port: parseInt(DATABASE_PORT, 10),
  user: DATABASE_USER,
  password: DATABASE_PASS,

  seeder: {
    path: './dist/seeders', // path to the folder with seeders
    pathTs: './seeders', // path to the folder with TS seeders (if used, we should put path to compiled files in `path`)
    defaultSeeder: 'DatabaseSeeder', // default seeder class name
    glob: '!(*.d).{js,ts}', // how to match seeder files (all .js and .ts files, but not .d.ts)
    emit: 'ts', // seeder generation mode
    fileName: (className: string) => className, // seeder file naming convention
  },
};

export default config;
```

Here's a Gist showing the logs: https://gist.github.com/fandy/93a6e8a9734214dd49c75a0ed46ae103

**Reply by B4nan** [reply]
So the very same queries, right? Sounds more like your problem is not they are not populated, but that they are not serialized that way.

**Reply by xndyz** [reply]
Is there any reason the serialization would differ between the two usages of `populate: true` vs `populate: ['chats.members']`?

**Reply by B4nan** [reply]
With `populate: true` there is no populate hint stored in the entity state (which is used when serializing the entity). Previously `populate: true` did generate full populate hint which is no longer needed for things to work, but probably has this side effect. Will see what we can do about this.

**Reply by B4nan** [reply]
I don't think we can do much here. There are recursion checks in place exactly for this use case. The serialization works based on populate hints exactly for this reason - so you are the one in control. With `populate: true` we don't know where to stop, so we stop when we see a cycle. `populate: true` was always targetted at prototyping and tests, not really as a regular thing to use in your app.

**Reply by xndyz** [reply]
Ah I see, thank you for the explanation! Sounds like I don't have to feel bad about specifying the populate fields individually.
